README.md

```diff
@@ -136,6 +136,27 @@ policy may be made to more closely align with other providers in the Cluster API
    You can find some guidance on what needs to be edited, and how to create some of the
    required OpenStack resources in the [Configuration documentation](docs/config.md).
 
+
+   #### Notes on configuring an APIServer LoadBalancer
+   
+   It's possible to configure a loadbalancer, which is then used to loadbalance in front of the control 
+   plane nodes. This feature is optional for single-node clusters but highly recommmended for multi-node 
+   control plane clusters. The loadbalancer can be configured via the following properties in the Cluster CRD:
+   
+   ````
+     managedAPIServerLoadBalancer: true
+     apiServerLoadBalancerFloatingIP: <floating-ip>
+     apiServerLoadBalancerPort: 6443
+     apiServerLoadBalancerAdditionalPorts:
+       - 22
```

Comment:
sorry, but I am little bit curious on 22, why we need a ssh to a LB? should this be ssh to a host only? 

And, I think if possible, can we create a template for non-LB and LB cluster.yaml template? (non-LB can reuse previuos one)

Reply:
We don't need it. But in my case I"m using it to ssh onto the nodes because I cannot access the Machines because I"m only setting a Floating IP on the LoadBalancer.

I would prefer adding these templates with multi-node control plane. I would say single-node control plane makes sense without lb and multi-node with lb.

To be honest I think we're loosing control over our permutations. We cannot really maintain & test all permutations of:
* networks are pre-created vs ClusterAPI creates them
* floating ips are pre-created vs ClusterAPI creates floating ips vs we're not using any
* using api server loadbalancer vs without api server loadbalancer
* single-node control plane vs multi-node control plane

I think we have to define our use cases from scratch and then document them cleanly. Right now I don't really now which permutations we support.

Reply:
I'll give it a try :):

## Networks:

* Networks on cluster-level
  * Specify nodeCIDR and network/subnet/router will be created and added to the Cluster status
  * [to be implemented] Specify networks (network/subnet) filters to detect existing networks and add them Cluster status

> Result is always that we have network/subnet in the cluster status

* Networks on machine-level
  * Per default the networks from the Cluster are used
  * It's possible to overwrite them by filter (I saw in the doc that it's possible to add the Machine to multiple networks)

## Floating IPs

* LoadBalancer
  * [to be implemented] Can be set optionally on the APIServer loadbalancer

* Machines
  * Can be set optionally on every machine
    * [to be implemented] create them if they are specified and don't exist

* APIServer IP
  * After multi-node control plane pr: the controlPlaneEndpoint has to match either:
    * [with loadbalancer]: the load balancer floating ip
      * what do we do if none is set? 
    * [without loadbalancer]: the ip of the single control plane machine
      * no lb and multiple control plane machines doesn't make sense imho

## LoadBalancer

* Loadbalancer can be used with single & multi-node control plane (allows easier update from single to multi-node control plane)
* Is Loadbalancer mandatory for multi-node control plane? I think yes, because if it isn't what can we use as controlPlaneEndpoint?

Reply:
> We don't need it. But in my case I"m using it to ssh onto the nodes because I cannot access the Machines because I"m only setting a Floating IP on the LoadBalancer.
> 
> I would prefer adding these templates with multi-node control plane. I would say single-node control plane makes sense without lb and multi-node with lb.
+1 to this, single node without lb and multi node with lb is the right way
> 
> To be honest I think we're loosing control over our permutations. We cannot really maintain & test all permutations of:
> 
> * networks are pre-created vs ClusterAPI creates them
this is because my env has some limitation, I can't use lb if it lead to mandatory nodecidr case, I think it's low priority in normal cases.....
> * floating ips are pre-created vs ClusterAPI creates floating ips vs we're not using any
currently the floating ips has to be created (either w/ or w/o LB, right) ? I think it's low priority too to support create floating ip on the fly
> * using api server loadbalancer vs without api server loadbalancer
I might lost some background, do we only have LB now ?without LB will make it single node case? 
> * single-node control plane vs multi-node control plane
we can't always request lb, so single node and multiple node support should be there
> * ...
> 
> I think we have to define our use cases from scratch and then document them cleanly. Right now I don't really now which permutations we support and how this dimensions interfere with each other.

Reply:
See the 2nd post from me above :)

Reply:
I totally agree non-lb case to sing-node ,this make sense :)

I can do the #428 thing, I am struggling with that now...

Reply:
Okay. Tell me if I can help you

Reply:
thanks a lot :), I am having 2 issues in my env:
1) I can't create floating ip ,I can only associate existing floating ip to LB 
2) the newly create network/subnet can't be used because of policy, I have to use existing network...

Reply:
> 
> 
> I'll give it a try :):
> ## Networks:
> 
>     * Networks on cluster-level
>       
>       * Specify nodeCIDR and network/subnet/router will be created and added to the Cluster status
>       * [to be implemented] Specify networks (network/subnet) filters to detect existing networks and add them Cluster status
> 
> 
> > Result is always that we have network/subnet in the cluster status
> 
>     * Networks on machine-level
>       
>       * Per default the networks from the Cluster are used (#425)
>       * It's possible to overwrite them by filter (I saw in the doc that it's possible to add the Machine to multiple networks)
> 
> 
> ## Floating IPs
> 
>     * LoadBalancer
>       
>       * [to be implemented] Can be set optionally on the APIServer loadbalancer
> 
>     * Machines
>       
>       * Can be set optionally on every machine
>         
>         * [to be implemented] create them if they are specified and don't exist
> 
>     * APIServer IP
>       
>       * After multi-node control plane pr: the controlPlaneEndpoint has to match either:
>         
>         * [with loadbalancer]: the load balancer floating ip
>           
>           * what do we do if none is set?
>         * [without loadbalancer]: the ip of the single control plane machine
>           
>           * no lb and multiple control plane machines doesn't make sense imho
> 
> 
> ## LoadBalancer
> 
>     * Loadbalancer can be used with single & multi-node control plane (allows easier update from single to multi-node control plane)
> 
>     * Is Loadbalancer mandatory for multi-node control plane? I think yes, because if it isn't what can we use as controlPlaneEndpoint?
> 
> 
> P.S.: If we have consensus an what makes sense and what we can support we should document these cases. I think this would lead to far better documentation to just document each permutation dimension (floating ip/network/ load balancer/ single vs multi-node control plane) by itself.

exactly, we have limitted resource so maybe we can focus on some case we think important first... we can add more support if we have to our priority...
to my case, as I indicated, I have to use existing network because my env limitation..

anyway, let's merge https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/421 first..